### PR TITLE
Check forbidden zone for typed deconstruction variables

### PR DESCRIPTION
**Pull request by alrz**
Fixes https://twitter.com/controlflow/status/1487757301461266434

Sounds like a bug. This probably needs approval as it's a breaking change.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs [hunk 1]
```diff
@@ -9722,6 +9722,27 @@ public void Deconstruct(out int a, out string b)
...
 
+        [Fact]
+        public void MixedDeclarationAndAssignmentForbiddenZone()
+        {
+            string source = @"
+class C
+{
+    static void Main()
+    {
+        (int x1, (x1, int x2), x2) = (1, (2, 3), 4);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:9733]: >(int x1, (x1, int x2), x2) = (1, (2, 3), 4); [](http://example.com/codeflow?start=8&length=44)

I see nothing wrong in this code and it looks like it is working as I would expect it to work.

**Comment by AlekseyTs** [comment]
I am going to close this PR since I have seen nothing that would support the claim that the current behavior is incorrect.

**Comment by 333fred** [comment]
I would agree with Aleksey. It maybe pointless, but I don't see any reason why this should be forbidden.

**Comment by alrz** [comment]
It feels like we're still on the "declaration part" of the deconstruction but I guess that's not how you look at it.

**Pull request closed without merging**